### PR TITLE
Fix ASOF LEFT JOIN performance degradation

### PR DESCRIPTION
**Pull request by yisheng-oai**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

address https://github.com/ClickHouse/ClickHouse/issues/46453. Bisect marked https://github.com/ClickHouse/ClickHouse/pull/35525 as the bad changed. This PR looks to reverse the changes in that PR.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by Algunenano** [comment]
Some comments:
* As expected [this makes some queries slower](https://s3.amazonaws.com/clickhouse-test-reports/47544/94ca9613a17e1f92be491139ec0594e7c6c1d6c1/performance_comparison_[1/4]/report.html)
* If you are going to do a change (or a revert) for performance reasons, it's better to tag them as performance and add a performance test with the affected query.
* This PR reverts way more than necessary, including some style changes that were correct when done. In the ticket I mentioned that just changing the storage class seemed to revert most of the slowdown, so I'd suggest keeping the changes to a minimum instead of a blind revert.

**Review by nickitat — CHANGES_REQUESTED** [review]
I agree with @Algunenano, issue is almost resolved [here](https://github.com/ClickHouse/ClickHouse/issues/46453#issuecomment-1432884753)

**Comment by ongkong** [comment]
@Algunenano @nickitat Great suggestions, i'll look into them. Could you give me a pointer to how [this performance comparison](https://s3.amazonaws.com/clickhouse-test-reports/47544/94ca9613a17e1f92be491139ec0594e7c6c1d6c1/performance_comparison_%5B1/4%5D/report.html) was generated in 
> Some comments:
> 
> * As expected [this makes some queries slower](https://s3.amazonaws.com/clickhouse-test-reports/47544/94ca9613a17e1f92be491139ec0594e7c6c1d6c1/performance_comparison_%5B1/4%5D/report.html)....

**Comment by Algunenano** [comment]
> Could you give me a pointer to how [this performance comparison](https://s3.amazonaws.com/clickhouse-test-reports/47544/94ca9613a17e1f92be491139ec0594e7c6c1d6c1/performance_comparison_%5B1/4%5D/report.html) 

It comes from https://github.com/ClickHouse/ClickHouse/blob/8e0a2a3a27a814cf5de250766d19d3a3696a5dc8/tests/performance/asof.xml which was added in a previous PR when working on performance for ASOF joins.

You can add new tests, with table creation and so on by simply creating new files. A good example would be something like [this file](https://github.com/ClickHouse/ClickHouse/blob/8e0a2a3a27a814cf5de250766d19d3a3696a5dc8/tests/performance/IPv4.xml). Just add a new .xml file with that format and the CI will pick it up and run performance tests with and without the changes (and in future PRs)

File: src/Interpreters/RowRefs.cpp [hunk 1]
```diff
@@ -75,8 +75,8 @@ class SortedLookupVector : public SortedLookupVectorBase
 
 public:
     using Keys = std::vector<TKey>;
-    using Entries = PaddedPODArray<Entry>;
-    using RowRefs = PaddedPODArray<RowRef>;
+    using Entries = std::vector<Entry>;
```
Comment (nickitat) [inline src/Interpreters/RowRefs.cpp:78]: I would replace both with something like `PODArrayWithStackMemory<Entry, sizeof(Entry)>`. maybe we could allocate a bit more on stack - would be good to benchmark.

File: src/Interpreters/RowRefs.cpp [hunk 2]
```diff
@@ -75,8 +75,8 @@ class SortedLookupVector : public SortedLookupVectorBase
 
 public:
     using Keys = std::vector<TKey>;
```
Comment (nickitat) [inline src/Interpreters/RowRefs.cpp:77]: let's remove

File: tests/performance/asof.xml [hunk 3]
```diff
@@ -41,4 +41,34 @@
         ON (visits.fingerprint = origins.fingerprint AND visits.date &gt;= origins.date)
         FORMAT Null
     </query>
+
+    <substitutions>
+        <substitution>
+            <name>num_rows</name>
+            <values>
+                <value>100000</value>
```
Comment (nickitat) [inline tests/performance/asof.xml:49]: 1'000'000 should also be fast enough. let's add it here too

File: src/Interpreters/RowRefs.cpp (hunk 1 above)
Comment (ongkong) [reply src/Interpreters/RowRefs.cpp]: I ran  `docker/test/performance-comparison/perf.py ./tests/performance/asof.xml` locally.
with `std::vector`:
```
prewarm	0	asof.query2.prewarm0	0	0.6126589775085449
query	0	asof.query2.run0	0	0.6231968402862549
query	0	asof.query2.run1	0	0.6166267395019531
```

with `PODArrayWithStackMemory `:
```
prewarm	0	asof.query2.prewarm0	0	0.33847498893737793
query	0	asof.query2.run0	0	0.3189249038696289
query	0	asof.query2.run1	0	0.31333398818969727
query	0	asof.query2.run2	0	0.31946492195129395
query	0	asof.query2.run3	0	0.3158740997314453
```

i pushed up the latest commit with `PODArrayWithStackMemory`

**Review by nickitat — APPROVED** [review]
lgtm

**Comment by nickitat** [comment]
perf tests look good. slowdowns are irrelevant and don't reproduce locally

**Comment by nickitat** [comment]
Stateless tests (tsan) [1/5] - #47460
Stateless tests (tsan, s3 storage) [3/5] - `DB::Exception: Too many parts` irrelevant

**Pull request merged**